Create CheckList for Safety Claims on a Generic Linux System.md (#28)
elisa-tech/wg-osep: Contributions/CheckList_for_Safety_Claims_on_a_Generic_Linux_System.md

[Structure of the document](#Structure-of-the-document)

[Hardware Platform](#Hardware-Platform)

[Toolchain](#Toolchain)

[Linux Safety Checklist](#Linux-Safety-Checklist)

[License: CC BY-SA 4.0](#License-CC-BY-SA-40)

## **Terms and Abbreviations**

   See the matching section in *Interference Scenarios for an ARM64 Linux System*

## **References**

1. ***Interference Scenarios for an ARM64 Linux System***
2. ***Safety Requirements for a Generic Linux System***
3. [CC BY-SA 4.0 Deed | Attribution-ShareAlike 4.0 International | Creative Commons](https://creativecommons.org/licenses/by-sa/4.0/) - <https://creativecommons.org/licenses/by-sa/4.0/> License

## **Purpose of the document**
This checklist is meant to help with preparing a Linux-based safety solution.

The challenge with using Linux, is that it is a very complex, ever-evolving, crowd-sourced system, and it doesn't follow the traditional safety processes and workflows.

Furthermore, unless typical off-the-shelf commercial components, it doesn't already come with a safety manual.

Even the architectural descriptions that can be found through documentation, books, courses, etc. are not created with safety in mind.

> The risk is that one might overlook aspects which are paramount to safety, but typically deemed irrelevant for linux users and even to kernel developers they might appear as very niche.

Reviewer: ```suggestion
The risk is that one might overlook aspects that are paramount to safety, but typically deemed irrelevant for Linux users, and even to kernel developers they might appear as very niche.
```

The list here provided (as-is) does not claim to be sufficient, nor correct (Linux evolves rapidly), and it is strongly advised to confirm that the statements made are still correct and relevant.

However, it is an attempt to bring the attention to aspects that might not be immediately obvious.
One immediate example of such an aspect is the fact that it is practically impossible to talk about safety without defining the reference hardware platform.

The entire document will mostly avoid entering in the details of both hardware implementation and Linux internals, beyond what is strictly required to justify each individual item of the checklist.
Similarly, it won't provide extensive justification of why something might be relevant, from a safety perspective.

> All of the background justification, related to hardware (ARM64), Linux kernel architecture, and allocation of safety requirements, they can be found in the documents mentioned in the [References](#References) section.

Reviewer: ```suggestion
All of the background justification, related to hardware (ARM64), Linux kernel architecture, and allocation of safety requirements, can be found in the documents mentioned in the [References](#References) section.
```

## **Structure of the document**
- The next two sections are a brief - but necessary - premise about hardware and toolchains: Linux is a code-base, but what gets deployed is a combination of hardware and binary code. This aspect cannot be overlooked.
- In the following section, the actual list unfolds.

## **Hardware Platform**

When a system is expected to meet certain safety requirements, these requirements are typically allocated to some high-level functionality, which is what constitutes the actual focus of interest in a safety context.

The system is then iteratively broken down into components, and safety requirements are percolated accordingly. At each iteration, the components having to support safety requirements may be further broken down, until the complexity of the components identified is deemed to be treatable.

At the very bottom of the stack of components lies the hardware. Hardware components may be qualified to a given safety level. Or not.

The level of safety qualification (or lack of) presented by the relevant hardware components can affect significantly the overall strategy used for the qualification of the entire system.

> For example, by using chips that have been both designed and produced following qualified processes, implemented with qualified tools, it is possible to make assumptions about hardware safety.

Reviewer: ```suggestion
For example, by using chips that have been both designed and produced following qualified processes, and implemented with qualified tools, it is reasonable to make assumptions about hardware safety.
```

Author: In this case I meant "possible", as in "legitimate". As objectie enabler. I feel that "reasonable" waters down the strenght I inentionaly wanted to use. Assuming that I'm attributing the correct meaning to theose terms.

 * *code injection*, to implement certain functionality transparently (some might say behind the back), for example:
   * *Control Flow Integrity*: security hardening feature that attempts to detect if a function is not being entered or exited through the designated entry/exit points. This is intended to thwart attacks that rely on the re-arrangement of the execution of instructions, through the localisation and exploitation of fragments of binary code (gadgets) that are leveraged out-of-order (ROP/JOP: Return-Oriented-Programming / Jump-Oriented-Programming)
   * *Stack Canaries*: to detect the accidental (or malicious) overwriting of the call stack, the compiler can inject code at the preamble/epilogue of a function, placing an additional value on the call stack, when invoking a function, and verifying at the exit that said value was not overwritten.
 * *speculative inlining* (typically customisable through hints), where the compiler can decide if/when to follow hints present in the code about performing replacement of code, rather than invoking a function. But it can also decide that some "regular" function is worth being inlined, even when it was not explicitly requested.

>  * *randomisation of structure layout*: as security measure, the compiler can alter the order of the fields of a structure, when it meets certain criteria (e.g. fields are exclusively pointers), because the only constraint impose by the C language on structures layout is that the address of a structure must coincide with the address of its first field. Altering the order of subsequent fields can thwart attempts of an attacker to alter specific memory locations, based on the expected content. The layout is randomised at compile time. However it means that a bug overwriting part of a structure can expose different behaviour, depending on what layout was picked by the compiler but a certain build round.

Reviewer: ```suggestion
 * *randomisation of structure layout*: as security measure, the compiler can alter the order of the fields of a structure, when it meets certain criteria (e.g. fields are exclusively pointers), because the only constraint impose by the C language on structure layout is that the address of a structure must coincide with the address of its first field. Altering the order of subsequent fields can thwart attempts of an attacker to alter specific memory locations, based on the expected content. The layout is randomised at compile time. However, this means that a bug overwriting part of a structure can expose different behaviour, depending on what layout was picked by the compiler for a certain build round.
```

 * *Link-Time-Optimisation*: the neat steps described in the toolchain flow above are altered in this way:
   * the compiler still generates assembly (as in the previous sequence)

>    * the assembler does not produce anymore object code, but rather a bitstream, retaining more information from the previous phase

Reviewer: ```suggestion
   * the assembler no longer produces object code, but rather a bitstream, retaining more information from the previous phase
```

These examples show how converting source code into instructions that can actually be executed by a processor is not a trivial step, and there are plenty of opportunities for the sequence to diverge from the nominal behaviour, making a case for the need to qualify the toolchain as well.
And because of the variance in the output of the very same toolchain, based on the options selected and the associated values, it is thus necessary for the toolchain to be used with the same parameters that were used during its qualification.

It is therefore necessary, as preliminary step, to verify that the toolchain is qualified accordingly to safety requirements, and used accordingly to its qualification. Lacking this, then it becomes necessary to identify an adequate mitigation strategy.

## **Linux Safety Checklist**

> The items presented below are not necessarily coupled with each other, however they are loosely ordered starting from the most fundamental ones, with the intention of creating a "base of safety" for each new one. This doesn't prevent a design where mitigation take a different route.

Reviewer: ```suggestion
The items presented below are not necessarily coupled with each other, but they are loosely ordered, starting from the most fundamental ones, with the intention of creating a "base of safety" for each new one. This doesn't prevent a design where mitigation takes a different route.
```

However, even if addressed through a fundamentally different approach, they should still be somehow addressed explicitly, even if only to say that they might not be applicable, or that they are handled otherwise.

The analysis covers safety woes, which need to be addressed when making any safety claims.

Safety woes belong primary to two classes:
* integrity - the safety-related functionality is in a corrupted state\

> In its basic form, integrity mandates the detection of corruption, within an allotted time, so that some pre-defined reaction can happen timely. Stricter integrity requirements would instead dictate that the corruption doesn't impair the functionality. 

Reviewer: ```suggestion
In its basic form, integrity mandates the detection of corruption, within an allotted time, so that some pre-defined reaction can happen in a timely way. Stricter integrity requirements would instead dictate that the corruption doesn't impair the functionality. 
```

* availability - the safety-related functionality becomes unavailable\

> Having any actual availability requirement will percolate in some form on the integrity requirements. High availability doesn't automatically imply that corruption cannot happen, but that it cannot cause interruption of a service. Restarting the affect component or having redundancy are examples of non-preventive mitigation, however any mitigation needs to be evaluated against the specific use-case at hand.

Reviewer: ```suggestion
Having any actual availability requirement will have an impact in some form on the integrity requirements. High availability doesn't automatically imply that corruption cannot happen, but that it cannot cause interruption of a service. Restarting the affected component or having redundancy are examples of non-preventive mitigation, however, any mitigation needs to be evaluated against the specific use-case at hand.
```

Some of the cases considered might appear as double-failure, which is usually ignored in the context of functional safety, however they are instead cascading failures stemming from a single one, and therefore fully relevant to this analysis.

**Items List:**

* [Homogeneity of the test/validation/verification environment](#Homogeneity-of-the-test/validation/verification-environment)
* [Intra-kernel spatial interference](#Intra-kernel-spatial-interference)
* [Inter-process interference for resources and side effects](#Inter-process-interference-for-resources-and-side-effects)

### **Homogeneity of the test/validation/verification environment**

> This list item is not teribly specific to Linux architecture, however it cannot be overlooked, because doing so can compromise the usefulness of any further work.

Reviewer: ```suggestion
This list item is not specific to Linux architecture, but it cannot be overlooked, because doing so can compromise the usefulness of any further work.
```

    * Even within the same revision, especially in the case or ARM processors, many hardware features are considered to be optional, or even customer-defined, or configurable. Caches, interrupt routing, presence of certain coprocessors, can (and will) be different, altering the flow of the execution.

  * **Discrete peripherals population**

    Variance in number, type or even make of peripherals can introduce uncontrolled changes in he behaviour of the system, due to difference in device drivers being loaded, event patterns, utilisation of system memory.
    
  * **Toolchain parameters**

    Some of the functionality mentioned above might be in different states of activation (certain options are not even binary; e.g. optimisation can have any of 4 values, typically)

  * **Kernel version**

    The fact that the instances of the OS being analysed are all Linux instances is a good start, but far from being sufficient. One needs to confirm that they are the same version, including additional patching that - in the case of distro vendors - might introduce unforeseen variance.

>     The truly reliable way to ensure consistency is to compare the SHA signature of the HEAD of the code bases, in case one is attempting to reuse existing measurements from the field. But it's fundamentally a risky approach, and in the signatures are different, then one should proof that the code bases are equivalent, effectively analysing the different patching levels.

Reviewer: ```suggestion
    The truly reliable way to ensure consistency is to compare the SHA signature of the HEAD of the code bases, in case one is attempting to reuse existing measurements from the field. But it's fundamentally a risky approach, and if the signatures are different, then one should prove that the code bases are equivalent, effectively analysing the different patching levels.
```

  * **Kernel configuration**

    Even after having confirmed that indeed, one is using exactly the same code base, the configuration options must be confirmed to be compatible. Once more, it's trivial to detect identity, but the complexity of proving equivalence between non-identical kernel configurations can escalate rapidly and soon become unmanageable.

  * **Applications**

>     Like if dealing with hardware configurations, toolchains setup and the kernel itself was not enough, applications too can interfere with measurements, in various ways and with differnet levels of subtlety:

Reviewer: ```suggestion
    As if dealing with hardware configurations, toolchains setup and the kernel itself was not enough, applications too can interfere with measurements, in various ways and with different levels of subtlety:
```

    *  the presence of different applications in user-space, even if unrelated to safety, can alter the behaviour of the system, for example altering memory pressure, network traffic, cpu load

>     *  the presence of monitoring applications can alter the power saving patterns; for example a particularly aggressive monitoring of system states can precent the system from entering into lower power states, leaving which would have greater latency. Once the monitoring application is removed, the system might be outside of acceptable operating margins.

Reviewer: ```suggestion
    *  the presence of monitoring applications can alter the power saving patterns; for example a particularly aggressive monitoring of system states can prevent the system from entering into lower power states, leaving which would have greater latency. Once the monitoring application is removed, the system might be outside of acceptable operating margins.
```

  * **Presence of monitoring equipment**

    Monitoring equipment - e.g. a serial port can have similar effects to the presence of a monitoring application:
    * detection of monitoring hardware (like a USB-to-serial FTDI adapter) can cause the loading of additional kernel modules
    * depending on the overall system configuration, some script might alter the logging level, causing heightened kernel activity and delays due to the higher traffic on the console.
    * should the logging exceed the log buffer capacity, it might even cause the notification of important events to be dropped
    * having to power additional Hardware peripherals might force power rails and clock sources to stay active, hiding issues related to their timely wakeup
    * even subtler issues: leaving monitoring options enabled but not active, like a disconnected serial port, might cause ground lines to be floating and generate spurious interrupts that, again, might hide latency problems due to the system being kept active.

### **Intra-kernel spatial interference**

> Premise: the linux kernel is monolithic, with the following design choices:

Reviewer: ```suggestion
Premise: the Linux kernel is monolithic, with the following design choices:
```

  
  * data used by device drivers, both for housekeeping and for their core functionality, including buffers\
  Can be altered in ways that are not noticeable (specific to each case).

  * data used by housekeeping components (e.g. scheduler, timers, memory managers, file systems, networking, pipes, etc.)

  * pages used for backing userspace processes\
    Can be any part of a process (code, rodata, stack, heap) and it can be accessed directly by the kernel, through the linear map; protections setup in the process page tables are completely irrelevant. A typical mistake is to assume that processes are somehow protected from the kernel, at least while they are not invoking syscalls. It couldn't be more wrong. Processes involved in safety scenarios are actually perhaps more exposed than non safe processes.\
    Why? because this type of interference acts on physical pages, and a victim process needs to have physical pages in use. Processes with safety requirements are likely to be marked as non swappable and have a higher residency in physical pages (ideally 100%), therefore becoming statistically more exposed.

> * A somewhat less obvious (and thus more vicious) type of interference is the cascaded one: a first component interferes with a second one one, which then misbehaves and affects a third, which was part of a safety scenario.\

Reviewer: ```suggestion
* A somewhat less obvious (and thus more vicious) type of interference is the cascaded one: a first component interferes with a second one, which then misbehaves and affects a third, which was part of a safety scenario.\
```
